services/venice-controller/src/main/java/com/linkedin/venice/controller/init/ControllerClientBackedSystemSchemaInitializer.java

```diff
@@ -29,16 +31,21 @@
 public class ControllerClientBackedSystemSchemaInitializer {
   private static final Logger LOGGER = LogManager.getLogger(ControllerClientBackedSystemSchemaInitializer.class);
   private static final String DEFAULT_KEY_SCHEMA_STR = "\"int\"";
-  private static final int DEFAULT_RETRY_TIMES = 10;
+  /**
+   * Current leader controller might transit from leader to standby, clear cached store repository, and fail to handle
+   * schema requests. We leverage retries to cover the leader->standby period so that a later retry will be sent to the
+   * new leader controller, who can handle schema requests successfully.
+   */
+  private static final int DEFAULT_RETRY_TIMES = 20;
 
   private final AvroProtocolDefinition protocolDefinition;
   private final String clusterName;
   private final VeniceHelixAdmin admin;
   private final Schema keySchema;
   private final UpdateStoreQueryParams storeMetadataUpdate;
   private final boolean autoRegisterPartialUpdateSchema;
-  private final boolean enforceSSLOnly;
-  private ControllerClient controllerClient;
+  private final VeniceControllerConfig controllerConfig;
+  private D2ControllerClient d2ControllerClient;
```

Comment:
Is the controller tightly-coupled to D2 with this change? Or can the `D2ControllerClient` connect to non-D2 URLs as well...? We have open source users that wish to not run D2, so it would be better to avoid introducing a tight coupling... can it be made to use one or the other?

Reply:
Thanks for the prompt review. Good call. I didn't know that there are open source users not willing to use D2. Are they planning to configure child.cluster.url instead?

If so, I can make code changes here to be able to construct controller client using child.cluster.url config. If not, there will be two issues:
1. Before initializing helix manager, we can not use VeniceHelixAdmin#getLeaderController to get leader url. Without controller url and d2, the feature that registers system schemas upon controller startup won't work for these users.
2. Parent controller construct child controller clients to query job status, get future version etc. Child controller client are built via url or d2 ([code](https://github.com/linkedin/venice/blob/23d91ddbadeba045764b71391056560a10e0740e/services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java#L1510)). If neither url nor d2 is available, parent controller can not query job or store info in child regions.